### PR TITLE
Task2

### PR DESCRIPTION
**Pull request by EvgeniyaSirenko**

### CODE REVIEW

File: src/main/java/com/epam/preprod/sirenko/SafeContainer.java [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+	public int size() {
+		return array.length;
+	}
+	
+	/**
+	 * Removes all the elements
+	 **/
+	@Override
+	public void clear() {
+		Object[] objects = Arrays.copyOf(array, array.length + 1);
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/SafeContainer.java]: We dont need it here.We can just create new array(what is done currently in 41 line)

File: src/main/java/com/epam/preprod/sirenko/SafeContainer.java [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+	 **/
+	@Override
+	public Iterator<E> iterator() {
+		return new SafeIterator<>(array, 0);
+	}
+	
+	/**
+	 * Iterates elements in this container in proper sequence
+	 **/
+	private class SafeIterator<T extends Product> implements Iterator<T> {
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/SafeContainer.java]: Think about better naming

File: src/main/java/com/epam/preprod/sirenko/SafeContainer.java [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+		private int index;
+		private final Object[] arrayCopy;
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/SafeContainer.java]: Same hare about name

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
+package com.epam.preprod.sirenko;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+public class TwoListWrapper<E extends Container> implements List<E>{
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]: We dont need to extend Container, we can use any collection here

File: src/main/java/com/epam/preprod/sirenko/SafeContainer.java [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+import com.epam.preprod.sirenko.entity.Product;
+
+import java.util.*;
+
+/**
+ * Resizable array-implementation of the List interface
+ *
+ * @author E.Sirenko
+ **/
+public class SafeContainer<E extends Product> implements List<E> {
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/SafeContainer.java]: Maybe we should use better naming to indicate behavior, for example CopyOnWriteContainer

File: src/main/test/com/epam/preprod/sirenko/SafeContainerTest.java [hunk 6]
```diff
@@ -0,0 +1,721 @@
...
+		
+		assertNotNull(container.get(1));
+		assertEquals(container.get(1), clothing);
+	}
+	
+	@Test
+	void shouldAddAllElements() {
+		SafeContainer<Product> container = new SafeContainer<>();
+		SafeContainer<Product> expectedContainer = new SafeContainer<>();
+		SafeContainer<Product> containerToAdd = new SafeContainer<>();
```
Comment (v-zahoruiko) [inline src/main/test/com/epam/preprod/sirenko/SafeContainerTest.java]: You dont need 3 containers here

File: src/main/test/com/epam/preprod/sirenko/SafeContainerTest.java [hunk 7]
```diff
@@ -0,0 +1,721 @@
...
+		container.set(0, food);
+		assertEquals(clothing, iterator.next());
+		assertEquals(food, container.get(0));
+	}
+	
+	@Test
+	void listIteratorShouldThrowUnsupportedOperationException() {
+		SafeContainer<Product> container = new SafeContainer<>();
+		
+		assertThrows(UnsupportedOperationException.class, () -> {
```
Comment (v-zahoruiko) [inline src/main/test/com/epam/preprod/sirenko/SafeContainerTest.java]: (minor)Can be replaced my method reference (look at sonar message)

File: src/main/test/com/epam/preprod/sirenko/SafeContainerTest.java [hunk 8]
```diff
@@ -0,0 +1,721 @@
...
+	@Test
+	void iteratorShouldUseUnchangedContainer() {
+		SafeContainer<Product> container = new SafeContainer<>();
+		Clothing clothing = new Clothing();
+		Food food = new Food();
+		DryFood dryFood = new DryFood();
+		container.add(clothing);
+		container.add(food);
+		container.add(dryFood);
+		Iterator<Product> iterator = container.iterator();
```
Comment (v-zahoruiko) [inline src/main/test/com/epam/preprod/sirenko/SafeContainerTest.java]: (minor)Please add spaces to better visibility(given when then pattern for tests)

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+	private List<E> unmodifiedList;
+	private List<E> modifiedList;
+	
+	
+	public TwoListWrapper(List unmodifiedList, List modifiedList) {
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]: add 'E' for lists

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+		}
+		return object;
+	}
+	
+	/**
+	 * Appends the specified element to the end of modified list
+	 */
+	@Override
+	public boolean add(E element) {
+		modifiedList.add(element);
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]:
```suggestion
		return modifiedList.add(element);
```

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+	}
+	
+	/**
+	 * Appends all of the elements in the specified collection to the end of
+	 * modified list, in the order that they are returned by the specified
+	 * collection's iterator (optional operation).
+	 */
+	@Override
+	public boolean addAll(Collection collection) {
+		modifiedList.addAll(collection);
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]:
```suggestion
		return modifiedList.addAll(collection);
```

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+	 * at the specified position (optional operation).  Shifts the
+	 * element currently at that position (if any) and any subsequent
+	 * elements to the right (increases their indices).
+	 */
+	@Override
+	public boolean addAll(int index, Collection collection) {
+		if (index < unmodifiedList.size()) {
+			throw new IllegalArgumentException("Can't add element to unmodified list");
+		} else {
+			modifiedList.addAll(index - unmodifiedList.size(), collection);
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]:
```suggestion
			return modifiedList.addAll(index - unmodifiedList.size(), collection);
```

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+	 * specified collection (optional operation).
+	 */
+	@Override
+	public boolean removeAll(Collection collection) {
+		for (Object o : collection) {
+			if (unmodifiedList.contains(o)) {
+				throw new IllegalArgumentException("Can't remove collection from unmodified list");
+			}
+		}
+		modifiedList.removeAll(collection);
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]:
```suggestion
		return modifiedList.removeAll(collection);
```

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+	 * specified collection (optional operation).  In other words, removes
+	 * from this list all of its elements that are not contained in the
+	 * specified collection.
+	 */
+	@Override
+	public boolean retainAll(Collection collection) {
+		if (!collection.containsAll(unmodifiedList)) {
+			throw new IllegalArgumentException("Can't retain collection of unmodified list");
+		} else {
+			modifiedList.retainAll(collection);
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]:
```suggestion
			return modifiedList.retainAll(collection);
```

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+	 * specified element (optional operation).
+	 */
+	@Override
+	public E set(int index, E element) {
+		if (index < unmodifiedList.size()) {
+			throw new IllegalArgumentException("Can't set element to unmodified list");
+		} else {
+			modifiedList.set(index - unmodifiedList.size(), element);
+		}
+		return modifiedList.get(index - unmodifiedList.size());
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]: can we use just return modifiedList.set(index - unmodifiedList.size(), element); here without calling get method?

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+	 * (optional operation).  Shifts the element currently at that position
+	 * (if any) and any subsequent elements to the right (adds one to their
+	 * indices).
+	 */
+	@Override
+	public void add(int index, E element) {
+		if (index < unmodifiedList.size()) {
+			throw new IllegalArgumentException("Can't add element to unmodified list");
+		} else {
+			modifiedList.add(index - unmodifiedList.size(), element);
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]: (minor)We can just use modifiedList.add(index - unmodifiedList.size(), element); without 'else' statement

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+	 * list.
+	 */
+	@Override
+	public E remove(int index) {
+		E toRemove;
+		if (index < unmodifiedList.size()) {
+			throw new IllegalArgumentException("Can't remove element to unmodified list");
+		} else {
+			toRemove = modifiedList.get(index - unmodifiedList.size());
+			modifiedList.remove(index - unmodifiedList.size());
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]:
```suggestion
			return modifiedList.remove(index - unmodifiedList.size());
```

File: src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+		return modifiedList.indexOf(object) + unmodifiedList.size();
+	}
+	
+	/**
+	 * Returns the index of the last occurrence of the specified element
+	 * in this list, or -1 if this list does not contain the element.
+	 */
+	@Override
+	public int lastIndexOf(Object object) {
+		if (unmodifiedList.contains(object)) {
```
Comment (v-zahoruiko) [inline src/main/java/com/epam/preprod/sirenko/TwoListWrapper.java]: unmodifiedList is before modified, but we looking for last index.So we need to check modifiedList first

File: src/main/test/com/epam/preprod/sirenko/TwoListWrapperTest.java [hunk 19]
```diff
@@ -0,0 +1,403 @@
...
+		modifiedList.add("!");
+		List<Object> listToRetain = new ArrayList<>();
+		listToRetain.add("fella");
+		listToRetain.add("Hello");
+		listToRetain.add(1);
+		listToRetain.add("good");
+		
+		twoListWrapper.retainAll(listToRetain);
+		
+		System.out.println(twoListWrapper.get(0));
```
Comment (v-zahoruiko) [inline src/main/test/com/epam/preprod/sirenko/TwoListWrapperTest.java]: We dont need souts it tests

Comment (EvgeniyaSirenko) [reply src/main/test/com/epam/preprod/sirenko/TwoListWrapperTest.java]: forgot to delete them

**Review by VitaliiZahoruiko — APPROVED** [review]

**Pull request merged**